src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs

```diff
@@ -0,0 +1,137 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.VisualStudio.CodingConventions;
+
+namespace Microsoft.CodeAnalysis.Editor.Options
+{
+    // NOTE: this type depends on Dev15 assemblies, which is why the type is in EditorFeatures.Next. But, that library
+    // is rehostable and once we move .editorconfig support fully through the system, it should be moved to Workspaces
+    // or perhaps even lower.
+    internal sealed class EditorConfigDocumentOptionsProvider : IDocumentOptionsProvider
+    {
+        private readonly object _gate = new object();
+
+        /// <summary>
+        /// The map of cached contexts for currently open documents.
+        /// </summary>
+        private readonly Dictionary<DocumentId, Task<ICodingConventionContext>> _openDocumentContexts = new Dictionary<DocumentId, Task<ICodingConventionContext>>();
```

Comment:
i assuem, because of the _gate, that this will be accesed from different threads.  can you doc as much?